File: frontend/app/api/process-youtube/ocr.js
import { createWorker, PSM } from "tesseract.js";
import path from "path";

import { deriveOcrIndicators, estimateImageSizeFromLines } from "./indicators";

const OCR_LANGUAGE = process.env.TESSERACT_LANGUAGE || "eng";
const MIN_WORD_CONFIDENCE = Number(process.env.TESSERACT_MIN_CONFIDENCE || 35);
const OCR_ENABLED = process.env.ENABLE_KEYFRAME_OCR !== "false";
const OCR_RECOGNITION_TIMEOUT_MS = Number(
  process.env.TESSERACT_RECOGNITION_TIMEOUT_MS || 8000,
);
const TESSERACT_ROOT = path.join(process.cwd(), "node_modules", "tesseract.js");
const TESSERACT_CORE_ROOT = path.join(
  process.cwd(),
  "node_modules",
  "tesseract.js-core",
);

function normalizeBox(box) {
  if (!box) {
    return null;
  }

  return {
    x0: box.x0,
    y0: box.y0,
    x1: box.x1,
    y1: box.y1,
  };
}

function collectLines(blocks) {
  const lines = [];

  for (const block of blocks || []) {
    for (const paragraph of block.paragraphs || []) {
      for (const line of paragraph.lines || []) {
        const text = line.text?.trim();
        if (!text || line.confidence < MIN_WORD_CONFIDENCE) {
          continue;
        }

        lines.push({
          text,
          confidence: line.confidence,
          boundingBox: normalizeBox(line.bbox),
        });
      }
    }
  }

  return lines;
}

function emptyOcrResult(error = null) {
  const lines = [];

  return {
    ok: !error,
    text: {
      raw: "",
      lines,
      wordCount: 0,
    },
    indicators: deriveOcrIndicators({ lines, raw: "" }),
    confidence: null,
    error,
  };
}

export async function createOcrProvider() {
  if (!OCR_ENABLED) {
    return disabledProvider("OCR disabled by ENABLE_KEYFRAME_OCR=false.");
  }

  const worker = await createWorker(OCR_LANGUAGE, 1, {
    workerPath: path.join(TESSERACT_ROOT, "src", "worker", "node", "index.js"),
    corePath: path.join(TESSERACT_CORE_ROOT, "tesseract-core-simd.wasm.js"),
  });
  await worker.setParameters({
    tessedit_pageseg_mode: PSM.SPARSE_TEXT,
  });

  return {
    async analyzeFrame(framePath) {
      try {
        const result = await withTimeout(
          worker.recognize(framePath),
          OCR_RECOGNITION_TIMEOUT_MS,
          `Tesseract OCR timed out after ${OCR_RECOGNITION_TIMEOUT_MS} ms.`,
        );
        const rawText = result.data.text?.trim() || "";
        const lines = collectLines(result.data.blocks);
        const { width, height } = estimateImageSizeFromLines(lines);
        const indicators = deriveOcrIndicators({
          lines,
          raw: rawText,
          width,
          height,
        });

        return {
          ok: true,
          text: {
            raw: rawText,
            lines,
            wordCount: lines.reduce(
              (count, line) => count + line.text.split(/\s+/).filter(Boolean).length,
              0,
            ),
          },
          indicators,
          confidence: result.data.confidence ?? null,
          error: null,
        };
      } catch (error) {
        return emptyOcrResult(error?.message || "OCR failed.");
      }
    },

    async close() {
      await worker.terminate();
    },
  };
}

function disabledProvider(message) {
  return {
    async analyzeFrame() {
      return emptyOcrResult(message);
    },

    async close() {},
  };
}

function withTimeout(promise, timeoutMs, message) {
  return Promise.race([
    promise,
    new Promise((_, reject) => {
      setTimeout(() => reject(new Error(message)), timeoutMs);
    }),
  ]);
}

File: backend/pipeline/ocr.py
from pathlib import Path

from PIL import Image

from .config import MIN_WORD_CONFIDENCE, OCR_ENABLED, OCR_LANGUAGE
from .indicators import derive_ocr_indicators, estimate_image_size_from_lines
from .tools import configure_tesseract, resolve_tesseract_path

configure_tesseract()

try:
    import pytesseract
except ImportError:
    pytesseract = None


def empty_ocr_result(error: str | None = None) -> dict:
    lines: list[dict] = []
    return {
        "ok": error is None,
        "text": {"raw": "", "lines": lines, "wordCount": 0},
        "indicators": derive_ocr_indicators(lines=lines, raw=""),
        "confidence": None,
        "error": error,
    }


def analyze_frame(frame_path: Path) -> dict:
    if not OCR_ENABLED:
        return empty_ocr_result("OCR disabled by ENABLE_KEYFRAME_OCR=false.")

    if pytesseract is None:
        return empty_ocr_result(
            "pytesseract is not installed. Install backend requirements and system tesseract-ocr."
        )

    if not resolve_tesseract_path():
        return empty_ocr_result(
            "Tesseract OCR binary not found. Install: winget install UB-Mannheim.TesseractOCR "
            "or set TESSERACT_CMD to tesseract.exe path."
        )

    try:
        image = Image.open(frame_path)
        data = pytesseract.image_to_data(
            image,
            lang=OCR_LANGUAGE,
            config="--psm 11",
            output_type=pytesseract.Output.DICT,
        )

        lines = _collect_lines(data)
        raw_text = _build_raw_text(lines)
        size = estimate_image_size_from_lines(lines)
        indicators = derive_ocr_indicators(
            lines=lines,
            raw=raw_text,
            width=size["width"],
            height=size["height"],
        )
        confidences = [line["confidence"] for line in lines]
        page_confidence = round(sum(confidences) / len(confidences), 2) if confidences else None

        return {
            "ok": True,
            "text": {
                "raw": raw_text,
                "lines": lines,
                "wordCount": sum(len(line["text"].split()) for line in lines),
            },
            "indicators": indicators,
            "confidence": page_confidence,
            "error": None,
        }
    except Exception as exc:
        return empty_ocr_result(str(exc))


def _collect_lines(data: dict) -> list[dict]:
    line_map: dict[tuple[int, int, int], dict] = {}

    count = len(data.get("text", []))
    for index in range(count):
        text = (data["text"][index] or "").strip()
        try:
            confidence = float(data["conf"][index])
        except (TypeError, ValueError):
            confidence = -1.0

        if not text or confidence < MIN_WORD_CONFIDENCE:
            continue

        key = (
            int(data["block_num"][index]),
            int(data["par_num"][index]),
            int(data["line_num"][index]),
        )
        left = int(data["left"][index])
        top = int(data["top"][index])
        width = int(data["width"][index])
        height = int(data["height"][index])
        box = {
            "x0": left,
            "y0": top,
            "x1": left + width,
            "y1": top + height,
        }

        if key not in line_map:
            line_map[key] = {
                "text": text,
                "confidence": confidence,
                "boundingBox": box,
            }
            continue

        entry = line_map[key]
        entry["text"] = f"{entry['text']} {text}".strip()
        entry["confidence"] = round((entry["confidence"] + confidence) / 2, 2)
        entry["boundingBox"]["x1"] = max(entry["boundingBox"]["x1"], box["x1"])
        entry["boundingBox"]["y1"] = max(entry["boundingBox"]["y1"], box["y1"])

    return sorted(line_map.values(), key=lambda item: (item["boundingBox"]["y0"], item["boundingBox"]["x0"]))


def _build_raw_text(lines: list[dict]) -> str:
    return "\n".join(line["text"] for line in lines).strip()
